[Bug]: after startup, first several search episodic memory fails

### Describe the bug

start memmachine:

2025-12-06 21:13:32,974 [10.2.100.33] [DEBUG pexpect_util.py::pexpect_interact::329::MainThread] pinter: send cmd: uv run bash memmachine-compose.sh
...
2025-12-06 21:14:14,216 [10.2.100.33] [DEBUG pexpect_util.py::pexpect_read::262::MainThread] pread: got a line(0):qa_idx=None: b'\x1b[0;32m[SUCCESS]\x1b[0m MemMachine is ready'
2025-12-06 21:14:14,217 [10.2.100.33] [DEBUG pexpect_util.py::pexpect_read::262::MainThread] pread: got a line(0):qa_idx=None: b'\x1b[0;32m[SUCCESS]\x1b[0m \xf0\x9f\x8e\x89 MemMachine is now running!'


GET /health, it returns healthy
run locomo test, console reports errors:

mmai_run_search:start time: 2025-12-06 21:16:20

mmai health:
k=status v=healthy
k=service v=memmachine
top_k=50
Processing questions for group 0...
^MAnswering Questions:   0%|                              | 0/152 [00:00<?, ?it/s]pc:ERROR: memory search failed ex=ERROR: status_code=500 reason=Internal Server Error
pc:Traceback (most recent call last):
  File "/home/memmachine/MemMachine/evaluation/locomo/episodic_memory/restapiv2_locomo_search.py", line 104, in process_question
    data = mmai.search_memory(question, top_k=top_k, types=types)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/memmachine/MemMachine/evaluation/locomo/utils/memmachine_helper_restapiv2.py", line 368, in search_memory
    raise AssertionError(f'ERROR: status_code={resp.status_code} reason={resp.reason}')
AssertionError: ERROR: status_code=500 reason=Internal Server Error
....
save: category=1 items=32 errors=2
save: category=2 items=37 errors=2
save: category=3 items=13 errors=1
save: category=4 items=70 errors=0

# the number of errors varies with each run


requests logs:

2025-12-06 21:16:27,468 [10.2.100.37] [DEBUG restapiv2_locomo_search.py::process_question::103::ThreadPoolExecutor-0_0] search memory conv=1 q=1
2025-12-06 21:16:27,469 [10.2.100.37] [DEBUG memmachine_helper_restapiv2.py::search_memory::361::ThreadPoolExecutor-0_0] POST url=http://localhost:8080/api/v2/memories/search payload={'top_k': 50, 'types': ['episodic'], 'query': 'When did Caroline go to
2025-12-06 21:16:27,472 [10.2.100.37] [DEBUG restapiv2_locomo_search.py::process_question::103::ThreadPoolExecutor-0_1] search memory conv=1 q=2
2025-12-06 21:16:27,472 [10.2.100.37] [DEBUG memmachine_helper_restapiv2.py::search_memory::361::ThreadPoolExecutor-0_1] POST url=http://localhost:8080/api/v2/memories/search payload={'top_k': 50, 'types': ['episodic'], 'query': 'When did Melanie paint
2025-12-06 21:16:27,475 [10.2.100.37] [DEBUG restapiv2_locomo_search.py::process_question::103::ThreadPoolExecutor-0_2] search memory conv=1 q=3
2025-12-06 21:16:27,475 [10.2.100.37] [DEBUG memmachine_helper_restapiv2.py::search_memory::361::ThreadPoolExecutor-0_2] POST url=http://localhost:8080/api/v2/memories/search payload={'top_k': 50, 'types': ['episodic'], 'query': 'What fields would Carol
2025-12-06 21:16:27,478 [10.2.100.37] [DEBUG restapiv2_locomo_search.py::process_question::103::ThreadPoolExecutor-0_3] search memory conv=1 q=4
2025-12-06 21:16:27,478 [10.2.100.37] [DEBUG memmachine_helper_restapiv2.py::search_memory::361::ThreadPoolExecutor-0_3] POST url=http://localhost:8080/api/v2/memories/search payload={'top_k': 50, 'types': ['episodic'], 'query': 'What did Caroline resea
2025-12-06 21:16:27,480 [10.2.100.37] [DEBUG restapiv2_locomo_search.py::process_question::103::ThreadPoolExecutor-0_4] search memory conv=1 q=5
2025-12-06 21:16:27,480 [10.2.100.37] [DEBUG memmachine_helper_restapiv2.py::search_memory::361::ThreadPoolExecutor-0_4] POST url=http://localhost:8080/api/v2/memories/search payload={'top_k': 50, 'types': ['episodic'], 'query': "What is Caroline's iden
2025-12-06 21:16:27,481 [10.2.100.37] [DEBUG restapiv2_locomo_search.py::process_question::103::ThreadPoolExecutor-0_5] search memory conv=1 q=6
2025-12-06 21:16:27,483 [10.2.100.37] [DEBUG restapiv2_locomo_search.py::process_question::103::ThreadPoolExecutor-0_6] search memory conv=1 q=7
2025-12-06 21:16:27,485 [10.2.100.37] [DEBUG restapiv2_locomo_search.py::process_question::103::ThreadPoolExecutor-0_7] search memory conv=1 q=8
2025-12-06 21:16:27,486 [10.2.100.37] [DEBUG memmachine_helper_restapiv2.py::search_memory::361::ThreadPoolExecutor-0_5] POST url=http://localhost:8080/api/v2/memories/search payload={'top_k': 50, 'types': ['episodic'], 'query': 'When did Melanie run a
2025-12-06 21:16:27,486 [10.2.100.37] [DEBUG restapiv2_locomo_search.py::process_question::103::ThreadPoolExecutor-0_8] search memory conv=1 q=9
2025-12-06 21:16:27,493 [10.2.100.37] [DEBUG memmachine_helper_restapiv2.py::search_memory::361::ThreadPoolExecutor-0_8] POST url=http://localhost:8080/api/v2/memories/search payload={'top_k': 50, 'types': ['episodic'], 'query': 'When did Caroline give
2025-12-06 21:16:27,487 [10.2.100.37] [DEBUG restapiv2_locomo_search.py::process_question::103::ThreadPoolExecutor-0_9] search memory conv=1 q=10
2025-12-06 21:16:27,494 [10.2.100.37] [DEBUG memmachine_helper_restapiv2.py::search_memory::361::ThreadPoolExecutor-0_9] POST url=http://localhost:8080/api/v2/memories/search payload={'top_k': 50, 'types': ['episodic'], 'query': 'When did Caroline meet
2025-12-06 21:16:27,487 [10.2.100.37] [DEBUG memmachine_helper_restapiv2.py::search_memory::361::ThreadPoolExecutor-0_6] POST url=http://localhost:8080/api/v2/memories/search payload={'top_k': 50, 'types': ['episodic'], 'query': 'When is Melanie plannin
2025-12-06 21:16:27,488 [10.2.100.37] [DEBUG memmachine_helper_restapiv2.py::search_memory::361::ThreadPoolExecutor-0_7] POST url=http://localhost:8080/api/v2/memories/search payload={'top_k': 50, 'types': ['episodic'], 'query': "What is Caroline's rela
2025-12-06 21:16:27,542 [10.2.100.37] [DEBUG memmachine_helper_restapiv2.py::search_memory::365::ThreadPoolExecutor-0_1] status=500 reason=Internal Server Error
2025-12-06 21:16:27,542 [10.2.100.37] [DEBUG memmachine_helper_restapiv2.py::search_memory::366::ThreadPoolExecutor-0_1] text=Internal Server Error
2025-12-06 21:16:27,544 [10.2.100.37] [DEBUG restapiv2_locomo_search.py::process_question::103::ThreadPoolExecutor-0_1] search memory conv=1 q=11
2025-12-06 21:16:27,545 [10.2.100.37] [DEBUG memmachine_helper_restapiv2.py::search_memory::361::ThreadPoolExecutor-0_1] POST url=http://localhost:8080/api/v2/memories/search payload={'top_k': 50, 'types': ['episodic'], 'query': 'How long has Caroline h
2025-12-06 21:16:27,556 [10.2.100.37] [DEBUG memmachine_helper_restapiv2.py::search_memory::365::ThreadPoolExecutor-0_2] status=500 reason=Internal Server Error
2025-12-06 21:16:27,556 [10.2.100.37] [DEBUG memmachine_helper_restapiv2.py::search_memory::366::ThreadPoolExecutor-0_2] text=Internal Server Error
2025-12-06 21:16:27,557 [10.2.100.37] [DEBUG restapiv2_locomo_search.py::process_question::103::ThreadPoolExecutor-0_2] search memory conv=1 q=12
2025-12-06 21:16:27,557 [10.2.100.37] [DEBUG memmachine_helper_restapiv2.py::search_memory::361::ThreadPoolExecutor-0_2] POST url=http://localhost:8080/api/v2/memories/search payload={'top_k': 50, 'types': ['episodic'], 'query': 'Where did Caroline move
2025-12-06 21:16:27,568 [10.2.100.37] [DEBUG memmachine_helper_restapiv2.py::search_memory::365::ThreadPoolExecutor-0_3] status=500 reason=Internal Server Error
2025-12-06 21:16:27,568 [10.2.100.37] [DEBUG memmachine_helper_restapiv2.py::search_memory::366::ThreadPoolExecutor-0_3] text=Internal Server Error
2025-12-06 21:16:27,570 [10.2.100.37] [DEBUG restapiv2_locomo_search.py::process_question::103::ThreadPoolExecutor-0_3] search memory conv=1 q=13
2025-12-06 21:16:27,570 [10.2.100.37] [DEBUG memmachine_helper_restapiv2.py::search_memory::361::ThreadPoolExecutor-0_3] POST url=http://localhost:8080/api/v2/memories/search payload={'top_k': 50, 'types': ['episodic'], 'query': "How long ago was Caroli
2025-12-06 21:16:27,584 [10.2.100.37] [DEBUG memmachine_helper_restapiv2.py::search_memory::365::ThreadPoolExecutor-0_4] status=500 reason=Internal Server Error
2025-12-06 21:16:27,585 [10.2.100.37] [DEBUG memmachine_helper_restapiv2.py::search_memory::366::ThreadPoolExecutor-0_4] text=Internal Server Error
2025-12-06 21:16:27,586 [10.2.100.37] [DEBUG restapiv2_locomo_search.py::process_question::103::ThreadPoolExecutor-0_4] search memory conv=1 q=14
2025-12-06 21:16:27,586 [10.2.100.37] [DEBUG memmachine_helper_restapiv2.py::search_memory::361::ThreadPoolExecutor-0_4] POST url=http://localhost:8080/api/v2/memories/search payload={'top_k': 50, 'types': ['episodic'], 'query': 'What career path has Ca
2025-12-06 21:16:27,615 [10.2.100.37] [DEBUG memmachine_helper_restapiv2.py::search_memory::365::ThreadPoolExecutor-0_0] status=500 reason=Internal Server Error
2025-12-06 21:16:27,615 [10.2.100.37] [DEBUG memmachine_helper_restapiv2.py::search_memory::366::ThreadPoolExecutor-0_0] text=Internal Server Error
2025-12-06 21:16:27,617 [10.2.100.37] [DEBUG restapiv2_locomo_search.py::process_question::103::ThreadPoolExecutor-0_0] search memory conv=1 q=15
2025-12-06 21:16:27,617 [10.2.100.37] [DEBUG memmachine_helper_restapiv2.py::search_memory::361::ThreadPoolExecutor-0_0] POST url=http://localhost:8080/api/v2/memories/search payload={'top_k': 50, 'types': ['episodic'], 'query': "Would Caroline still wa
2025-12-06 21:16:33,380 [10.2.100.37] [DEBUG memmachine_helper_restapiv2.py::search_memory::365::ThreadPoolExecutor-0_2] status=200 reason=OK
2025-12-06 21:16:33,380 [10.2.100.37] [DEBUG memmachine_helper_restapiv2.py::search_memory::366::ThreadPoolExecutor-0_2] text={"status":0,"content":{"episodic_memory":{"long_term_memory":{"episodes":[{"content":"Hey Mel! Good to see you! How have you be
2025-12-06 21:16:34,638 [10.2.100.37] [DEBUG memmachine_helper_restapiv2.py::search_memory::365::ThreadPoolExecutor-0_3] status=200 reason=OK


docker logs stdout:

2025-12-06 20:14:12,964 [INFO] alembic.runtime.migration - Context impl PostgresqlImpl.
2025-12-06 20:14:12,964 [INFO] alembic.runtime.migration - Will assume transactional DDL.
TOM1:_build_openai_embedders:obj=<memmachine.common.embedder.openai_embedder.OpenAIEmbedder object at 0x7f89ea9c4fe0> params={'client': <openai.AsyncOpenAI object at 0x7f89eaa6bda0>, 'model': 'text-embedding-3-small', 'dimensions': 1536, 'max_retry_
TOM1:_build_openai_chat_completions_language_model:obj=<memmachine.common.language_model.openai_chat_completions_language_model.OpenAIChatCompletionsLanguageModel object at 0x7f89ea9c5fa0> conf={'metrics_factory_id': 'prometheus', 'user_metrics_labe
2025-12-06 20:14:13,087 [INFO] mcp.server.streamable_http_manager - StreamableHTTP session manager started
INFO:     172.18.0.1:45162 - "GET /api/v2/health HTTP/1.1" 200 OK
2025-12-06 20:14:15,241 [INFO] httpx - HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"
2025-12-06 20:14:15,786 [INFO] httpx - HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"
2025-12-06 20:14:16,443 [INFO] httpx - HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"
INFO:     127.0.0.1:43790 - "GET /api/v2/health HTTP/1.1" 200 OK
2025-12-06 20:14:17,688 [INFO] httpx - HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"
2025-12-06 20:14:19,079 [INFO] httpx - HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"
....
2025-12-06 20:16:25,946 [INFO] httpx - HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"
2025-12-06 20:16:26,132 [INFO] httpx - HTTP Request: POST https://api.openai.com/v1/embeddings "HTTP/1.1 200 OK"
INFO:     172.18.0.1:34044 - "GET /api/v2/health HTTP/1.1" 200 OK
INFO:     172.18.0.1:34058 - "GET /api/v2/metrics HTTP/1.1" 200 OK
INFO:     172.18.0.1:34086 - "POST /api/v2/memories/search HTTP/1.1" 500 Internal Server Error
INFO:     172.18.0.1:34094 - "POST /api/v2/memories/search HTTP/1.1" 500 Internal Server Error
INFO:     172.18.0.1:34098 - "POST /api/v2/memories/search HTTP/1.1" 500 Internal Server Error
INFO:     172.18.0.1:34112 - "POST /api/v2/memories/search HTTP/1.1" 500 Internal Server Error
INFO:     172.18.0.1:34074 - "POST /api/v2/memories/search HTTP/1.1" 500 Internal Server Error
TOM1:_build_identity_reranker:name=id_ranker_id obj=<memmachine.common.reranker.identity_reranker.IdentityReranker object at 0x7f89db60da60>
2025-12-06 20:16:27,881 [INFO] httpx - HTTP Request: POST https://api.openai.com/v1/embeddings "HTTP/1.1 200 OK"
2025-12-06 20:16:27,890 [INFO] httpx - HTTP Request: POST https://api.openai.com/v1/embeddings "HTTP/1.1 200 OK"

docker logs stderr:
see attachment

[epi_mem_docker_log_search_errors.txt](https://github.com/user-attachments/files/23997709/epi_mem_docker_log_search_errors.txt)

### Steps to reproduce

see description above

### Expected behavior

search memory returns successfully

### Environment

rocky9
memmachine main/head 12/02 commit a0ac2a52ca16


### Additional context

_No response_

--- a/src/memmachine/common/resource_manager/resource_manager.py
+++ b/src/memmachine/common/resource_manager/resource_manager.py
@@ -2,6 +2,7 @@
 
 import asyncio
 import logging
+from asyncio import Lock
 
 from neo4j import AsyncDriver
 from sqlalchemy.ext.asyncio import AsyncEngine
@@ -65,6 +66,11 @@ class ResourceManagerImpl:
         self._episode_storage: EpisodeStorage | None = None
         self._semantic_manager: SemanticResourceManager | None = None
 
+        self._session_data_manager_lock = Lock()
+        self._episodic_memory_manager_lock = Lock()
+        self._episode_storage_lock = Lock()
+        self._semantic_manager_lock = Lock()
+
     async def build(self) -> None:
         """Build all configured resources in parallel."""
         tasks = [
@@ -125,47 +131,51 @@ class ResourceManagerImpl:
 
     async def get_session_data_manager(self) -> SessionDataManager:
         """Lazy-load the session data manager."""
-        if self._session_data_manager is not None:
-            return self._session_data_manager
-        database = self._conf.session_manager.database
-        engine = await self.get_sql_engine(database)
+        if self._session_data_manager is None:
+            async with self._session_data_manager_lock:
+                if self._session_data_manager is None:
+                    database = self._conf.session_manager.database
+                    engine = await self.get_sql_engine(database)
 
-        self._session_data_manager = SessionDataManagerSQL(engine)
-        await self._session_data_manager.create_tables()
-
+                    self._session_data_manager = SessionDataManagerSQL(engine)
+                    await self._session_data_manager.create_tables()
+        assert self._session_data_manager is not None
         return self._session_data_manager
 
     async def get_episodic_memory_manager(self) -> EpisodicMemoryManager:
         """Lazy-load the episodic memory manager."""
-        if self._episodic_memory_manager is not None:
-            return self._episodic_memory_manager
-        session_data_manager = await self.get_session_data_manager()
-        params = EpisodicMemoryManagerParams(
-            resource_manager=self,
-            session_data_manager=session_data_manager,
-        )
-        self._episodic_memory_manager = EpisodicMemoryManager(params)
+        if self._episodic_memory_manager is None:
+            async with self._episodic_memory_manager_lock:
+                if self._episodic_memory_manager is None:
+                    session_data_manager = await self.get_session_data_manager()
+                    params = EpisodicMemoryManagerParams(
+                        resource_manager=self,
+                        session_data_manager=session_data_manager,
+                    )
+                    self._episodic_memory_manager = EpisodicMemoryManager(params)
+        assert self._episodic_memory_manager is not None
         return self._episodic_memory_manager
 
     async def get_episode_storage(self) -> EpisodeStorage:
         """Return the episode storage instance."""
-        if self._episode_storage is not None:
-            return self._episode_storage
+        if self._episode_storage is None:
+            async with self._episode_storage_lock:
+                if self._episode_storage is None:
+                    episode_storage_conf = getattr(self._conf, "episode_storage", None)
+                    if episode_storage_conf is None:
+                        episode_storage_conf = self._conf.episode_store
 
-        episode_storage_conf = getattr(self._conf, "episode_storage", None)
-        if episode_storage_conf is None:
-            episode_storage_conf = self._conf.episode_store
+                    database = episode_storage_conf.database
+                    engine = await self.get_sql_engine(database)
 
-        database = episode_storage_conf.database
-        engine = await self.get_sql_engine(database)
+                    episode_storage = SqlAlchemyEpisodeStore(engine)
+                    await episode_storage.startup()
 
-        episode_storage = SqlAlchemyEpisodeStore(engine)
-        await episode_storage.startup()
+                    if episode_storage_conf.with_count_cache:
+                        episode_storage = CountCachingEpisodeStorage(episode_storage)
 
-        if episode_storage_conf.with_count_cache:
-            episode_storage = CountCachingEpisodeStorage(episode_storage)
-
-        self._episode_storage = episode_storage
+                    self._episode_storage = episode_storage
+        assert self._episode_storage is not None
         return self._episode_storage
 
     async def get_semantic_service(self) -> SemanticService:
@@ -175,16 +185,17 @@ class ResourceManagerImpl:
 
     async def get_semantic_manager(self) -> SemanticResourceManager:
         """Return the semantic resource manager, constructing if needed."""
-        if self._semantic_manager is not None:
-            return self._semantic_manager
-
-        episode_storage = await self.get_episode_storage()
-        self._semantic_manager = SemanticResourceManager(
-            semantic_conf=self._conf.semantic_memory,
-            prompt_conf=self._conf.prompt,
-            resource_manager=self,  # type: ignore[arg-type]
-            episode_storage=episode_storage,
-        )
+        if self._semantic_manager is None:
+            async with self._semantic_manager_lock:
+                if self._semantic_manager is None:
+                    episode_storage = await self.get_episode_storage()
+                    self._semantic_manager = SemanticResourceManager(
+                        semantic_conf=self._conf.semantic_memory,
+                        prompt_conf=self._conf.prompt,
+                        resource_manager=self,  # type: ignore[arg-type]
+                        episode_storage=episode_storage,
+                    )
+        assert self._semantic_manager is not None
         return self._semantic_manager
 
     async def get_semantic_session_manager(self) -> SemanticSessionManager:

